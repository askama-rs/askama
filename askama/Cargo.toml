[package]
name = "askama"
version = "0.13.0"
description = "Type-safe, compiled Jinja-like templates for Rust"
keywords = ["markup", "template", "jinja2", "html"]
categories = ["template-engine"]
homepage = "https://askama.readthedocs.io/"
repository = "https://github.com/askama-rs/askama"
license = "MIT OR Apache-2.0"
edition = "2021"
rust-version = "1.81"

[package.metadata.docs.rs]
features = ["full"]
rustdoc-args = ["--generate-link-to-definition", "--cfg=docsrs"]

[[bench]]
name = "to-json"
harness = false
required-features = ["serde_json"]

[[bench]]
name = "escape"
harness = false

[dependencies]
itoa = "1.0.11"

# needed by feature "derive"
askama_derive = { version = "=0.13.0", path = "../askama_derive", default-features = false, optional = true }

# needed by feature "serde_json"
serde = { version = "1.0", optional = true, default-features = false }
serde_json = { version = "1.0", optional = true, default-features = false }

# needed by feature "urlencode"
percent-encoding = { version = "2.1.0", optional = true, default-features = false }

[dev-dependencies]
assert_matches = "1.5.0"
criterion = "0.5"

[badges]
maintenance = { status = "actively-developed" }

[features]
default = ["config", "derive", "std", "urlencode", "askama_derive?/default"]
full = ["default", "blocks", "code-in-doc", "serde_json", "askama_derive?/full"]

alloc = [
    "askama_derive?/alloc",
    "serde?/alloc",
    "serde_json?/alloc",
    "percent-encoding?/alloc",
]
blocks = ["askama_derive?/blocks"]
code-in-doc = ["askama_derive?/code-in-doc"]
config = ["askama_derive?/config"]
derive = ["askama_derive"]
serde_json = ["std", "askama_derive?/serde_json", "dep:serde", "dep:serde_json"]
std = [
    "alloc",
    "askama_derive?/std",
    "serde?/std",
    "serde_json?/std",
    "percent-encoding?/std",
]
urlencode = ["askama_derive?/urlencode", "dep:percent-encoding"]
