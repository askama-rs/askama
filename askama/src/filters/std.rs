use core::convert::Infallible;
use std::collections::HashMap;
use std::collections::hash_map::Entry;
use std::hash::Hash;
use std::rc::Rc;

/// Returns an iterator with all duplicates removed.
///
/// The sorting order is kept, no extra allocation is performed. However, to make it possible,
/// the data is wrapped inside `Rc`.
///
/// ```
/// # use askama::Template;
/// #[derive(Template)]
/// #[template(ext = "html", source = "{% for elem in example|unique %}{{ elem }},{% endfor %}")]
/// struct Example<'a> {
///     example: Vec<&'a str>,
/// }
///
/// assert_eq!(
///     Example { example: vec!["a", "b", "a", "c"] }.to_string(),
///     "a,b,c,"
/// );
/// ```
pub fn unique<T: Hash + Eq>(
    it: impl IntoIterator<Item = T>,
) -> Result<impl Iterator<Item = Rc<T>>, Infallible> {
    let mut set = HashMap::new();

    Ok(it.into_iter().filter_map(move |elem| {
        // To prevent cloning the data, we need to use `Rc`, like that we can clone `elem` as
        // key of the `HashSet` and return it.
        if let Entry::Vacant(entry) = set.entry(Rc::new(elem)) {
            Some(Rc::clone(entry.insert_entry(()).key()))
        } else {
            None
        }
    }))
}

#[cfg(test)]
mod test {
    use alloc::vec; // It's the macro, not the module.
    use alloc::vec::Vec;

    use super::*;

    #[test]
    fn test_unique() {
        assert_eq!(
            unique(["a", "b", "a", "c"]).unwrap().collect::<Vec<_>>(),
            vec![Rc::new("a"), Rc::new("b"), Rc::new("c")]
        );
        assert_eq!(
            unique([1, 1, 1, 2, 1]).unwrap().collect::<Vec<_>>(),
            vec![Rc::new(1), Rc::new(2)]
        );
        assert_eq!(
            unique("hello".chars()).unwrap().collect::<Vec<_>>(),
            vec![Rc::new('h'), Rc::new('e'), Rc::new('l'), Rc::new('o')]
        );
    }
}
