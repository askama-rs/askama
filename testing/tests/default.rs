use std::borrow::Cow;
use std::num::{NonZeroU32, Saturating, Wrapping};
use std::pin::Pin;
use std::sync::Arc;

use askama::Template;

#[test]
fn test_default() {
    // In this test the filter |default should render the variable `value` if it is default.
    // Otherwise the default value "unknown" is rendered.

    #[derive(Template)]
    #[template(source = r#"{{ value | default("unknown") }}"#, ext = "html")]
    struct SuchVariable {
        value: i64,
    }

    assert_eq!(SuchVariable { value: 42 }.render().unwrap(), "42");

    #[derive(Template)]
    #[template(source = r#"{{ value | default("unknown") }}"#, ext = "html")]
    struct NoSuchVariable {
        _value: i64,
    }

    assert_eq!(NoSuchVariable { _value: 42 }.render().unwrap(), "unknown");

    #[derive(Template)]
    #[template(source = r#"{{ value | default("unknown", false) }}"#, ext = "html")]
    struct SuchVariable2 {
        value: i64,
    }

    assert_eq!(SuchVariable2 { value: 42 }.render().unwrap(), "42");

    #[derive(Template)]
    #[template(source = r#"{{ value | default("unknown", false) }}"#, ext = "html")]
    struct NoSuchVariable2 {
        _value: i64,
    }

    assert_eq!(NoSuchVariable2 { _value: 42 }.render().unwrap(), "unknown");
}

#[test]
fn test_defined_or() {
    // In this test the filter |defined_or should render the variable `value` if it is default.
    // Otherwise the default value "unknown" is rendered.

    #[derive(Template)]
    #[template(source = r#"{{ value | defined_or("unknown") }}"#, ext = "html")]
    struct SuchVariable {
        value: i64,
    }

    assert_eq!(SuchVariable { value: 42 }.render().unwrap(), "42");

    #[derive(Template)]
    #[template(source = r#"{{ value | defined_or("unknown") }}"#, ext = "html")]
    struct NoSuchVariable {
        _value: i64,
    }

    assert_eq!(NoSuchVariable { _value: 42 }.render().unwrap(), "unknown");
}

#[test]
fn test_default_var_with_unwrap() {
    // In this test the filter |default should try to unwrap the variable `value`.
    // If the variable exists and the value could be unwrapped, then the value is rendered.
    // Otherwise the default value "unknown" is rendered.

    #[derive(Template)]
    #[template(source = r#"{{ value | default("unknown", true) }}"#, ext = "html")]
    struct AnOption {
        value: Option<i64>,
    }

    assert_eq!(AnOption { value: None }.render().unwrap(), "unknown");
    assert_eq!(AnOption { value: Some(42) }.render().unwrap(), "42");

    #[derive(Template)]
    #[template(source = r#"{{ value | default("unknown", true) }}"#, ext = "html")]
    struct AResult {
        value: Result<i64, std::fmt::Error>,
    }

    assert_eq!(
        AResult {
            value: Err(std::fmt::Error)
        }
        .render()
        .unwrap(),
        "unknown"
    );
    assert_eq!(AResult { value: Ok(42) }.render().unwrap(), "42");

    #[derive(Template)]
    #[template(source = r#"{{ value | default("unknown", true) }}"#, ext = "html")]
    struct NoSuchVariable {
        _value: Option<i64>,
    }

    assert_eq!(NoSuchVariable { _value: None }.render().unwrap(), "unknown");
    assert_eq!(
        NoSuchVariable { _value: Some(42) }.render().unwrap(),
        "unknown"
    );
}

#[test]
fn test_assigned_or_var_with_unwrap() {
    // In this test the filter |assigned_or should try to unwrap the variable `value`.
    // If the variable exists and the value could be unwrapped, then the value is rendered.
    // Otherwise the default value "unknown" is rendered.

    #[derive(Template)]
    #[template(source = r#"{{ value | assigned_or("unknown") }}"#, ext = "html")]
    struct AnOption {
        value: Option<i64>,
    }

    assert_eq!(AnOption { value: None }.render().unwrap(), "unknown");
    assert_eq!(AnOption { value: Some(42) }.render().unwrap(), "42");

    #[derive(Template)]
    #[template(source = r#"{{ value | assigned_or("unknown") }}"#, ext = "html")]
    struct AResult {
        value: Result<i64, std::fmt::Error>,
    }

    assert_eq!(
        AResult {
            value: Err(std::fmt::Error)
        }
        .render()
        .unwrap(),
        "unknown"
    );
    assert_eq!(AResult { value: Ok(42) }.render().unwrap(), "42");

    #[derive(Template)]
    #[template(source = r#"{{ value | assigned_or("unknown") }}"#, ext = "html")]
    struct NoSuchVariable {
        _value: Option<i64>,
    }

    assert_eq!(NoSuchVariable { _value: None }.render().unwrap(), "unknown");
    assert_eq!(
        NoSuchVariable { _value: Some(42) }.render().unwrap(),
        "unknown"
    );
}

#[test]
fn test_default_expr_with_unwrap() {
    // In this test the filter |default should try to unwrap the outcome of an expression.
    // If the outcome could be unwrapped, then the value is rendered.
    // Otherwise the default value "unknown" is rendered.

    #[derive(Template)]
    #[template(
        source = r#"{{ NonZeroU32::new(*value) | default("unknown", true) }}"#,
        ext = "html"
    )]
    struct AnOption {
        value: u32,
    }

    assert_eq!(AnOption { value: 0 }.render().unwrap(), "unknown");
    assert_eq!(AnOption { value: 42 }.render().unwrap(), "42");

    #[derive(Template)]
    #[template(
        source = r#"{{ std::str::from_utf8(value) | default("unknown", true) }}"#,
        ext = "html"
    )]
    struct AResult<'a> {
        value: &'a [u8],
    }

    assert_eq!(AResult { value: b"H\xffllo" }.render().unwrap(), "unknown");
    assert_eq!(AResult { value: b"Hello" }.render().unwrap(), "Hello");
}

#[test]
fn test_assigned_or_expr_with_unwrap() {
    // In this test the filter |assigned_or should try to unwrap the outcome of an expression.
    // If the outcome could be unwrapped, then the value is rendered.
    // Otherwise the default value "unknown" is rendered.

    #[derive(Template)]
    #[template(
        source = r#"{{ NonZeroU32::new(*value) | assigned_or("unknown") }}"#,
        ext = "html"
    )]
    struct AnOption {
        value: u32,
    }

    assert_eq!(AnOption { value: 0 }.render().unwrap(), "unknown");
    assert_eq!(AnOption { value: 42 }.render().unwrap(), "42");

    #[derive(Template)]
    #[template(
        source = r#"{{ std::str::from_utf8(value) | assigned_or("unknown") }}"#,
        ext = "html"
    )]
    struct AResult<'a> {
        value: &'a [u8],
    }

    assert_eq!(AResult { value: b"H\xffllo" }.render().unwrap(), "unknown");
    assert_eq!(AResult { value: b"Hello" }.render().unwrap(), "Hello");
}

#[test]
fn test_default_with_int() {
    // In this test a default value should be returned for an integer value 0.

    #[derive(Template)]
    #[template(source = r#"{{ value | default("unknown", true) }}"#, ext = "html")]
    struct U8 {
        value: Arc<Pin<Box<u8>>>,
    }

    assert_eq!(
        U8 {
            value: Arc::new(Box::pin(0))
        }
        .render()
        .unwrap(),
        "unknown"
    );
    assert_eq!(
        U8 {
            value: Arc::new(Box::pin(42))
        }
        .render()
        .unwrap(),
        "42"
    );

    #[derive(Template)]
    #[template(source = r#"{{ value | default("unknown", true) }}"#, ext = "html")]
    struct WrappingU8 {
        value: Arc<Pin<Box<Wrapping<u8>>>>,
    }

    assert_eq!(
        WrappingU8 {
            value: Arc::new(Box::pin(Wrapping(0)))
        }
        .render()
        .unwrap(),
        "unknown"
    );
    assert_eq!(
        WrappingU8 {
            value: Arc::new(Box::pin(Wrapping(42)))
        }
        .render()
        .unwrap(),
        "42"
    );

    #[derive(Template)]
    #[template(source = r#"{{ value | default("unknown", true) }}"#, ext = "html")]
    struct SaturatingU8 {
        value: Arc<Pin<Box<Saturating<u8>>>>,
    }

    assert_eq!(
        SaturatingU8 {
            value: Arc::new(Box::pin(Saturating(0)))
        }
        .render()
        .unwrap(),
        "unknown"
    );
    assert_eq!(
        SaturatingU8 {
            value: Arc::new(Box::pin(Saturating(42)))
        }
        .render()
        .unwrap(),
        "42"
    );

    #[derive(Template)]
    #[template(source = r#"{{ value | default("unknown", true) }}"#, ext = "html")]
    struct NzU32 {
        value: NonZeroU32,
    }

    assert_eq!(
        NzU32 {
            value: NonZeroU32::new(42).unwrap()
        }
        .render()
        .unwrap(),
        "42"
    );
}

#[test]
fn test_assigned_or_with_int() {
    // In this test a default value should be returned for an integer value 0.

    #[derive(Template)]
    #[template(source = r#"{{ value | assigned_or("unknown") }}"#, ext = "html")]
    struct U8 {
        value: Arc<Pin<Box<u8>>>,
    }

    assert_eq!(
        U8 {
            value: Arc::new(Box::pin(0))
        }
        .render()
        .unwrap(),
        "unknown"
    );
    assert_eq!(
        U8 {
            value: Arc::new(Box::pin(42))
        }
        .render()
        .unwrap(),
        "42"
    );

    #[derive(Template)]
    #[template(source = r#"{{ value | assigned_or("unknown") }}"#, ext = "html")]
    struct WrappingU8 {
        value: Arc<Pin<Box<Wrapping<u8>>>>,
    }

    assert_eq!(
        WrappingU8 {
            value: Arc::new(Box::pin(Wrapping(0)))
        }
        .render()
        .unwrap(),
        "unknown"
    );
    assert_eq!(
        WrappingU8 {
            value: Arc::new(Box::pin(Wrapping(42)))
        }
        .render()
        .unwrap(),
        "42"
    );

    #[derive(Template)]
    #[template(source = r#"{{ value | assigned_or("unknown") }}"#, ext = "html")]
    struct SaturatingU8 {
        value: Arc<Pin<Box<Saturating<u8>>>>,
    }

    assert_eq!(
        SaturatingU8 {
            value: Arc::new(Box::pin(Saturating(0)))
        }
        .render()
        .unwrap(),
        "unknown"
    );
    assert_eq!(
        SaturatingU8 {
            value: Arc::new(Box::pin(Saturating(42)))
        }
        .render()
        .unwrap(),
        "42"
    );

    #[derive(Template)]
    #[template(source = r#"{{ value | assigned_or("unknown") }}"#, ext = "html")]
    struct NzU32 {
        value: NonZeroU32,
    }

    assert_eq!(
        NzU32 {
            value: NonZeroU32::new(42).unwrap()
        }
        .render()
        .unwrap(),
        "42"
    );
}

#[test]
fn test_default_with_str() {
    // In this test a default value should be returned for an empty string.

    #[derive(Template)]
    #[template(source = r#"{{ value | default("unknown", true) }}"#, ext = "html")]
    struct Str<'a> {
        value: &'a str,
    }

    assert_eq!(Str { value: "" }.render().unwrap(), "unknown");
    assert_eq!(Str { value: "Hello" }.render().unwrap(), "Hello");

    #[derive(Template)]
    #[template(source = r#"{{ value | default("unknown", true) }}"#, ext = "html")]
    struct AString {
        value: String,
    }

    assert_eq!(
        AString {
            value: "".to_owned()
        }
        .render()
        .unwrap(),
        "unknown"
    );
    assert_eq!(
        AString {
            value: "Hello".to_owned()
        }
        .render()
        .unwrap(),
        "Hello"
    );

    #[derive(Template)]
    #[template(source = r#"{{ value | default("unknown", true) }}"#, ext = "html")]
    struct ACow<'a> {
        value: Cow<'a, str>,
    }

    assert_eq!(ACow { value: "".into() }.render().unwrap(), "unknown");
    assert_eq!(
        ACow {
            value: "Hello".into()
        }
        .render()
        .unwrap(),
        "Hello"
    );
}

#[test]
fn test_assigned_or_with_str() {
    // In this test a default value should be returned for an empty string.

    #[derive(Template)]
    #[template(source = r#"{{ value | assigned_or("unknown") }}"#, ext = "html")]
    struct Str<'a> {
        value: &'a str,
    }

    assert_eq!(Str { value: "" }.render().unwrap(), "unknown");
    assert_eq!(Str { value: "Hello" }.render().unwrap(), "Hello");

    #[derive(Template)]
    #[template(source = r#"{{ value | assigned_or("unknown") }}"#, ext = "html")]
    struct AString {
        value: String,
    }

    assert_eq!(
        AString {
            value: "".to_owned()
        }
        .render()
        .unwrap(),
        "unknown"
    );
    assert_eq!(
        AString {
            value: "Hello".to_owned()
        }
        .render()
        .unwrap(),
        "Hello"
    );

    #[derive(Template)]
    #[template(source = r#"{{ value | assigned_or("unknown") }}"#, ext = "html")]
    struct ACow<'a> {
        value: Cow<'a, str>,
    }

    assert_eq!(ACow { value: "".into() }.render().unwrap(), "unknown");
    assert_eq!(
        ACow {
            value: "Hello".into()
        }
        .render()
        .unwrap(),
        "Hello"
    );
}

#[test]
fn test_default_with_bool() {
    // In this test a default value should be returned for `false`.

    #[derive(Template)]
    #[template(source = r#"{{ value | default("unknown", true) }}"#, ext = "html")]
    struct Bool {
        value: bool,
    }

    assert_eq!(Bool { value: false }.render().unwrap(), "unknown");
    assert_eq!(Bool { value: true }.render().unwrap(), "true");
}

#[test]
fn test_assigned_or_with_bool() {
    // In this test a default value should be returned for `false`.

    #[derive(Template)]
    #[template(source = r#"{{ value | assigned_or("unknown") }}"#, ext = "html")]
    struct Bool {
        value: bool,
    }

    assert_eq!(Bool { value: false }.render().unwrap(), "unknown");
    assert_eq!(Bool { value: true }.render().unwrap(), "true");
}

#[test]
fn test_default_with_forward_declaration() {
    // In this test, it is ensured that a forward declared variable gets replaced with a
    // default value.

    #[derive(Template)]
    #[template(
        source = "\
        {% let var %}\
        {{ var | default(\"unknown\") }}\
        {% if true %}{% let var = 42 %}{% endif %}",
        ext = "html"
    )]
    struct Fwd;

    assert_eq!(Fwd.render().unwrap(), "unknown");

    #[derive(Template)]
    #[template(
        source = "\
        {% let var %}{# shadowing happens here #}\
        {{ var | default(\"unknown\") }}\
        {% if true %}{% let var = 42 %}{% endif %}",
        ext = "html"
    )]
    struct Fwd2 {
        #[allow(dead_code)]
        var: u32,
    }

    assert_eq!(Fwd2 { var: 42 }.render().unwrap(), "unknown");
}

#[test]
fn test_defined_or_with_forward_declaration() {
    // In this test, it is ensured that a forward declared variable gets replaced with a
    // default value.

    #[derive(Template)]
    #[template(
        source = "\
        {% let var %}\
        {{ var | defined_or(\"unknown\") }}\
        {% if true %}{% let var = 42 %}{% endif %}",
        ext = "html"
    )]
    struct Fwd;

    assert_eq!(Fwd.render().unwrap(), "unknown");

    #[derive(Template)]
    #[template(
        source = "\
        {% let var %}{# shadowing happens here #}\
        {{ var | defined_or(\"unknown\") }}\
        {% if true %}{% let var = 42 %}{% endif %}",
        ext = "html"
    )]
    struct Fwd2 {
        #[allow(dead_code)]
        var: u32,
    }

    assert_eq!(Fwd2 { var: 42 }.render().unwrap(), "unknown");
}
