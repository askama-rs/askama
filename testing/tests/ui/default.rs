use std::sync::Mutex;

use askama::Template;

#[derive(Template)]
#[template(ext = "html", source = "{{ (1 + 1) | default(2) }}")]
struct DefaultLhsExpr;

#[derive(Template)]
#[template(ext = "html", source = "{{ var | default }}")]
struct DefaultNoDefaultValue;

#[derive(Template)]
#[template(ext = "html", source = "{{ var | default(boolean = true) }}")]
struct DefaultNoDefaultValue2;

#[derive(Template)]
#[template(ext = "html", source = "{{ value | default(2, true) }}")]
struct DefaultNotUnwrappable {
    value: Mutex<u32>,
}

#[derive(Template)]
#[template(ext = "html", source = "{{ var | default(1, 2) }}")]
struct DefaultNotABool;

#[derive(Template)]
#[template(ext = "html", source = "{{ var | default(1, boolean = 2) }}")]
struct DefaultNotABool2;

#[derive(Template)]
#[template(ext = "html", source = "{{ var | assigned_or }}")]
struct AssignedOrNoDefaultValue;

#[derive(Template)]
#[template(ext = "html", source = "{{ (1 + 1) | defined_or(2) }}")]
struct DefinedOrLhsExpr;

#[derive(Template)]
#[template(ext = "html", source = "{{ var | defined_or }}")]
struct DefinedOrNoDefaultValue;

fn main() {}
