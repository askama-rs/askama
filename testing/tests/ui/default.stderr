error: the `default` filter requires a variable name on its left-hand side
 --> DefaultLhsExpr.html:1:11
       "| default(2) }}"
 --> tests/ui/default.rs:6:35
  |
6 | #[template(ext = "html", source = "{{ (1 + 1) | default(2) }}")]
  |                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: `default_value` argument is missing when calling `default` filter; its arguments are: (default_value, [boolean])
 --> DefaultNoDefaultValue.html:1:7
       "| default }}"
  --> tests/ui/default.rs:10:35
   |
10 | #[template(ext = "html", source = "{{ var | default }}")]
   |                                   ^^^^^^^^^^^^^^^^^^^^^

error: `default_value` argument is missing when calling `default` filter; its arguments are: (default_value, [boolean])
 --> DefaultNoDefaultValue2.html:1:7
       "| default(boolean = true) }}"
  --> tests/ui/default.rs:14:35
   |
14 | #[template(ext = "html", source = "{{ var | default(boolean = true) }}")]
   |                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: the `default` filter takes a boolean literal as its optional second argument
 --> DefaultNotABool.html:1:20
       "2) }}"
  --> tests/ui/default.rs:24:35
   |
24 | #[template(ext = "html", source = "{{ var | default(1, 2) }}")]
   |                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: the `default` filter takes a boolean literal as its optional second argument
 --> DefaultNotABool2.html:1:30
       "2) }}"
  --> tests/ui/default.rs:28:35
   |
28 | #[template(ext = "html", source = "{{ var | default(1, boolean = 2) }}")]
   |                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: `fallback` argument is missing when calling `assigned_or` filter; its arguments are: (fallback)
 --> AssignedOrNoDefaultValue.html:1:7
       "| assigned_or }}"
  --> tests/ui/default.rs:32:35
   |
32 | #[template(ext = "html", source = "{{ var | assigned_or }}")]
   |                                   ^^^^^^^^^^^^^^^^^^^^^^^^^

error: the `defined_or` filter requires a variable name on its left-hand side
 --> DefinedOrLhsExpr.html:1:11
       "| defined_or(2) }}"
  --> tests/ui/default.rs:36:35
   |
36 | #[template(ext = "html", source = "{{ (1 + 1) | defined_or(2) }}")]
   |                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: `fallback` argument is missing when calling `defined_or` filter; its arguments are: (fallback)
 --> DefinedOrNoDefaultValue.html:1:7
       "| defined_or }}"
  --> tests/ui/default.rs:40:35
   |
40 | #[template(ext = "html", source = "{{ var | defined_or }}")]
   |                                   ^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: `Mutex<u32>` is not `|assigned_or` filterable
  --> tests/ui/default.rs:17:10
   |
17 | #[derive(Template)]
   |          ^^^^^^^^ `Mutex<u32>` is not `|assigned_or` filterable
   |
   = help: the trait `DefaultFilterable` is not implemented for `Mutex<u32>`
   = help: the following other types implement trait `DefaultFilterable`:
             &T
             &mut T
             Arc<T>
             Box<T>
             Cow<'_, T>
             MutexGuard<'_, T>
             NonZero<i128>
             NonZero<i16>
           and $N others
   = note: required for `&Mutex<u32>` to implement `DefaultFilterable`
note: required by a bound in `assigned_or`
  --> $WORKSPACE/askama/src/filters/default.rs
   |
   | pub fn assigned_or<L: DefaultFilterable, R>(
   |                       ^^^^^^^^^^^^^^^^^ required by this bound in `assigned_or`
   = note: this error originates in the derive macro `Template` (in Nightly builds, run with -Z macro-backtrace for more info)
